File: schemas.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator


class StrictBaseModel(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)


class ReportSourceModel(StrictBaseModel):
    title: str = Field(min_length=1)
    url: str = ""


class SearchResultModel(StrictBaseModel):
    tool_name: str = Field(min_length=1)
    title: str = Field(min_length=1)
    url: str = ""
    snippet: str = Field(min_length=1, max_length=700)
    score: float = Field(ge=0.0, le=1.0)
    source_type: str = Field(min_length=1)
    chunk_id: str = Field(min_length=1)


class SelectedEvidenceModel(SearchResultModel):
    pass


GuardrailStatus = Literal["ready", "needs_clarification", "blocked"]
GuardrailAction = Literal["proceed", "revise", "block"]
GuardrailTool = Literal["tavily", "wikipedia", "weather"]
GuardrailRiskFlag = Literal[
    "prompt_injection",
    "secret_exfiltration",
    "policy_bypass",
    "overscoped_request",
    "non_research_request",
]


class GuardrailEvaluationModel(StrictBaseModel):
    status: GuardrailStatus
    recommended_action: GuardrailAction
    warnings: list[str] = Field(default_factory=list, max_length=6)
    risk_flags: list[GuardrailRiskFlag] = Field(default_factory=list, max_length=5)
    allowed_tools: list[GuardrailTool] = Field(default_factory=lambda: ["tavily", "wikipedia", "weather"], min_length=1, max_length=3)
    explanation: str = Field(min_length=1, max_length=240)
    clarifying_question: str = Field(default="", max_length=240)


class GuardrailStateModel(StrictBaseModel):
    sanitized_question: str = Field(min_length=3, max_length=600)
    status: GuardrailStatus
    recommended_action: GuardrailAction
    warnings: list[str] = Field(default_factory=list, max_length=6)
    risk_flags: list[GuardrailRiskFlag] = Field(default_factory=list, max_length=5)
    allowed_tools: list[GuardrailTool] = Field(default_factory=lambda: ["tavily", "wikipedia", "weather"], min_length=1, max_length=3)
    explanation: str = Field(min_length=1, max_length=240)
    clarifying_question: str = Field(default="", max_length=240)


class RunMetricsModel(StrictBaseModel):
    iterations_used: int = Field(ge=0)
    evidence_items: int = Field(ge=0)
    unique_sources: int = Field(ge=0)
    history_candidates: int = Field(ge=0)
    retrieval_strategy: str = Field(min_length=1)
    rerank_applied: int = Field(ge=0, le=1)
    rerank_candidates: int = Field(ge=0)
    rerank_duplicates_removed: int = Field(ge=0)
    rerank_trimmed_for_limit: int = Field(ge=0)
    rerank_distinct_sources: int = Field(ge=0)


class DraftReportModel(StrictBaseModel):
    title: str = Field(min_length=1)
    findings: list[str] = Field(default_factory=list, min_length=3, max_length=6)
    sources: list[ReportSourceModel] = Field(default_factory=list)
    confidence: float = Field(ge=0.0, le=1.0)
    summary: str = Field(min_length=1, max_length=240)


class FinalReportModel(StrictBaseModel):
    title: str = Field(min_length=1)
    summary: str = Field(min_length=1, max_length=240)
    key_findings: list[str] = Field(default_factory=list)
    sources: list[ReportSourceModel] = Field(default_factory=list)
    confidence: float = Field(ge=0.0, le=1.0)
    published_report: str = Field(min_length=1, max_length=3200)


class PastTopicRecordModel(StrictBaseModel):
    question: str = Field(min_length=1)
    report: FinalReportModel
    user_id: str = Field(min_length=1)
    created_at: str = Field(min_length=1)


class HistoryReferenceModel(StrictBaseModel):
    question: str = Field(min_length=1)
    created_at: str = Field(min_length=1)


class HistoryReviewModel(StrictBaseModel):
    match_type: Literal["similar", "related", "new"]
    rationale: str = Field(min_length=1)
    relevant_history: list[HistoryReferenceModel] = Field(default_factory=list, max_length=3)


class ResearchPlanModel(StrictBaseModel):
    queries: list[str] = Field(min_length=2, max_length=3)


class ReasonerDecisionModel(StrictBaseModel):
    decision: Literal["CONTINUE", "DONE"]
    reason: str = Field(min_length=1)


class PublishedAnswerModel(StrictBaseModel):
    published_report: str = Field(min_length=1, max_length=3200)


class HistoryMatchModel(StrictBaseModel):
    question: str = Field(min_length=1)
    published_report: str = ""
    title: str = ""
    summary: str = ""
    user_id: str = ""
    created_at: str = ""


class HistoryInterruptModel(StrictBaseModel):
    action: Literal["review_history_match"]
    current_question: str = Field(min_length=1)
    match_type: Literal["similar", "related", "new"]
    rationale: str = Field(min_length=1)
    matches: list[HistoryMatchModel] = Field(default_factory=list)
    reuse_allowed: bool = False
    reuse_candidate: HistoryMatchModel | None = None


class EvidenceSelectionInterruptModel(StrictBaseModel):
    action: Literal["select_evidence_for_report"]
    question: str = Field(min_length=1)
    research_plan: list[str] = Field(default_factory=list)
    current_evidence: list[SearchResultModel] = Field(default_factory=list)
    instructions: str = Field(min_length=1)


class ReviewInterruptModel(StrictBaseModel):
    action: Literal["review_before_publish"]
    question: str = Field(min_length=1)
    iterations: int = Field(ge=0)
    draft: DraftReportModel


class CreateSessionResponse(StrictBaseModel):
    thread_id: str = Field(min_length=1)


class TranslateTextRequest(StrictBaseModel):
    text: str = Field(min_length=1, max_length=3200)
    target_language: str = Field(min_length=2, max_length=16)


class TranslateTextResponse(StrictBaseModel):
    translated_text: str = Field(min_length=1, max_length=5000)
    target_language: str = Field(min_length=2, max_length=16)


class StartResearchRequest(StrictBaseModel):
    thread_id: str = Field(min_length=1)
    question: str
    user_id: str = Field(min_length=1)
    max_iterations: int = Field(default=2, ge=1, le=6)

    @field_validator("question")
    @classmethod
    def validate_question_length(cls, value: str) -> str:
        question = value.strip()
        if len(question) < 8:
            raise ValueError("Question is too short. Please enter at least 8 characters.")
        if len(question) > 600:
            raise ValueError("Question is too long. Please keep it under 600 characters.")
        return question


class ResumeResearchRequest(StrictBaseModel):
    thread_id: str = Field(min_length=1)
    decision: Literal[
        "proceed_with_context",
        "start_fresh_plan",
        "reuse_existing",
        "selected_evidence",
        "approved",
        "edited",
        "rejected",
    ]
    selected_evidence_ids: list[str] = Field(default_factory=list, max_length=12)
    human_feedback: str = Field(default="", max_length=800)


class RunSnapshotResponse(StrictBaseModel):
    thread_id: str = Field(min_length=1)
    status: Literal["idle", "waiting_input", "completed"]
    question: str = ""
    user_id: str = ""
    max_iterations: int = Field(default=0, ge=0)
    research_plan: list[str] = Field(default_factory=list)
    history_decision: str = ""
    review_decision: str = ""
    guardrails: GuardrailStateModel | None = None
    run_metrics: RunMetricsModel | None = None
    interrupt: HistoryInterruptModel | EvidenceSelectionInterruptModel | ReviewInterruptModel | None = None
    draft_report: DraftReportModel | None = None
    search_results: list[SearchResultModel] = Field(default_factory=list)
    selected_evidence_ids: list[str] = Field(default_factory=list)
    selected_evidence: list[SelectedEvidenceModel] = Field(default_factory=list)
    final_report: FinalReportModel | None = None
    reused_topic: PastTopicRecordModel | None = None

File: react-frontend/src/schemas.js
const nonEmptyString = { type: 'string', minLength: 1 }

const reportSourceSchema = {
  type: 'object',
  additionalProperties: false,
  required: ['title', 'url'],
  properties: {
    title: nonEmptyString,
    url: { type: 'string' },
  },
}

const searchResultSchema = {
  type: 'object',
  additionalProperties: false,
  required: ['tool_name', 'title', 'url', 'snippet', 'score', 'source_type', 'chunk_id'],
  properties: {
    tool_name: nonEmptyString,
    title: nonEmptyString,
    url: { type: 'string' },
    snippet: nonEmptyString,
    full_snippet: { type: 'string' },
    score: { type: 'number', minimum: 0, maximum: 1 },
    source_type: nonEmptyString,
    chunk_id: nonEmptyString,
  },
}

const selectedEvidenceSchema = searchResultSchema

const guardrailStateSchema = {
  type: 'object',
  additionalProperties: false,
  required: ['sanitized_question', 'status', 'recommended_action', 'warnings', 'risk_flags', 'allowed_tools', 'explanation', 'clarifying_question'],
  properties: {
    sanitized_question: nonEmptyString,
    status: { enum: ['ready', 'needs_clarification', 'blocked'] },
    recommended_action: { enum: ['proceed', 'revise', 'block'] },
    warnings: { type: 'array', items: { type: 'string' } },
    risk_flags: {
      type: 'array',
      items: {
        enum: ['prompt_injection', 'secret_exfiltration', 'policy_bypass', 'overscoped_request', 'non_research_request'],
      },
    },
    allowed_tools: {
      type: 'array',
      items: { enum: ['tavily', 'wikipedia'] },
      minItems: 1,
      maxItems: 2,
    },
    explanation: nonEmptyString,
    clarifying_question: { type: 'string' },
  },
}

const runMetricsSchema = {
  type: 'object',
  additionalProperties: false,
  required: ['iterations_used', 'evidence_items', 'unique_sources', 'history_candidates', 'retrieval_strategy', 'rerank_applied', 'rerank_candidates', 'rerank_duplicates_removed', 'rerank_trimmed_for_limit', 'rerank_distinct_sources'],
  properties: {
    iterations_used: { type: 'integer', minimum: 0 },
    evidence_items: { type: 'integer', minimum: 0 },
    unique_sources: { type: 'integer', minimum: 0 },
    history_candidates: { type: 'integer', minimum: 0 },
    retrieval_strategy: nonEmptyString,
    rerank_applied: { type: 'integer', minimum: 0, maximum: 1 },
    rerank_candidates: { type: 'integer', minimum: 0 },
    rerank_duplicates_removed: { type: 'integer', minimum: 0 },
    rerank_trimmed_for_limit: { type: 'integer', minimum: 0 },
    rerank_distinct_sources: { type: 'integer', minimum: 0 },
  },
}

const draftReportSchema = {
  type: 'object',
  additionalProperties: false,
  required: ['title', 'findings', 'sources', 'confidence', 'summary'],
  properties: {
    title: nonEmptyString,
    findings: { type: 'array', items: { type: 'string' } },
    sources: { type: 'array', items: reportSourceSchema },
    confidence: { type: 'number', minimum: 0, maximum: 1 },
    summary: nonEmptyString,
  },
}

const finalReportSchema = {
  type: 'object',
  additionalProperties: false,
  required: ['title', 'summary', 'key_findings', 'sources', 'confidence', 'published_report'],
  properties: {
    title: nonEmptyString,
    summary: nonEmptyString,
    key_findings: { type: 'array', items: { type: 'string' } },
    sources: { type: 'array', items: reportSourceSchema },
    confidence: { type: 'number', minimum: 0, maximum: 1 },
    published_report: nonEmptyString,
  },
}

const pastTopicSchema = {
  type: 'object',
  additionalProperties: false,
  required: ['question', 'report', 'user_id', 'created_at'],
  properties: {
    question: nonEmptyString,
    report: finalReportSchema,
    user_id: nonEmptyString,
    created_at: nonEmptyString,
  },
}

const historyMatchSchema = {
  type: 'object',
  additionalProperties: false,
  required: ['question', 'published_report', 'title', 'summary', 'user_id', 'created_at'],
  properties: {
    question: nonEmptyString,
    published_report: { type: 'string' },
    title: { type: 'string' },
    summary: { type: 'string' },
    user_id: { type: 'string' },
    created_at: { type: 'string' },
  },
}

const historyInterruptSchema = {
  type: 'object',
  additionalProperties: false,
  required: ['action', 'current_question', 'match_type', 'rationale', 'matches', 'reuse_allowed', 'reuse_candidate'],
  properties: {
    action: { const: 'review_history_match' },
    current_question: nonEmptyString,
    match_type: { enum: ['similar', 'related', 'new'] },
    rationale: nonEmptyString,
    matches: { type: 'array', items: historyMatchSchema },
    reuse_allowed: { type: 'boolean' },
    reuse_candidate: { anyOf: [historyMatchSchema, { type: 'null' }] },
  },
}

const evidenceSelectionInterruptSchema = {
  type: 'object',
  additionalProperties: false,
  required: ['action', 'question', 'research_plan', 'current_evidence', 'instructions'],
  properties: {
    action: { const: 'select_evidence_for_report' },
    question: nonEmptyString,
    research_plan: { type: 'array', items: { type: 'string' } },
    current_evidence: { type: 'array', items: searchResultSchema },
    instructions: nonEmptyString,
  },
}

const reviewInterruptSchema = {
  type: 'object',
  additionalProperties: false,
  required: ['action', 'question', 'iterations', 'draft'],
  properties: {
    action: { const: 'review_before_publish' },
    question: nonEmptyString,
    iterations: { type: 'integer', minimum: 0 },
    draft: draftReportSchema,
  },
}

export const createSessionResponseSchema = {
  type: 'object',
  additionalProperties: false,
  required: ['thread_id'],
  properties: {
    thread_id: nonEmptyString,
  },
}

export const startResearchRequestSchema = {
  type: 'object',
  additionalProperties: false,
  required: ['thread_id', 'question', 'user_id', 'max_iterations'],
  properties: {
    thread_id: nonEmptyString,
    question: { type: 'string', minLength: 8, maxLength: 600 },
    user_id: nonEmptyString,
    max_iterations: { type: 'integer', minimum: 1, maximum: 6 },
  },
}

export const resumeResearchRequestSchema = {
  type: 'object',
  additionalProperties: false,
  required: ['thread_id', 'decision', 'human_feedback'],
  properties: {
    thread_id: nonEmptyString,
    decision: {
      enum: [
        'proceed_with_context',
        'start_fresh_plan',
        'reuse_existing',
        'selected_evidence',
        'approved',
        'edited',
        'rejected',
      ],
    },
    selected_evidence_ids: { type: 'array', items: { type: 'string' } },
    human_feedback: { type: 'string', maxLength: 800 },
  },
}

export const runSnapshotResponseSchema = {
  type: 'object',
  additionalProperties: false,
  required: [
    'thread_id',
    'status',
    'question',
    'user_id',
    'max_iterations',
    'research_plan',
    'history_decision',
    'review_decision',
    'guardrails',
    'run_metrics',
    'interrupt',
    'draft_report',
    'search_results',
    'selected_evidence_ids',
    'selected_evidence',
    'final_report',
    'reused_topic',
  ],
  properties: {
    thread_id: nonEmptyString,
    status: { enum: ['idle', 'waiting_input', 'completed'] },
    question: { type: 'string' },
    user_id: { type: 'string' },
    max_iterations: { type: 'integer', minimum: 0 },
    research_plan: { type: 'array', items: { type: 'string' } },
    history_decision: { type: 'string' },
    review_decision: { type: 'string' },
    guardrails: {
      anyOf: [
        { type: 'null' },
        guardrailStateSchema,
      ],
    },
    run_metrics: {
      anyOf: [
        { type: 'null' },
        runMetricsSchema,
      ],
    },
    interrupt: {
      anyOf: [
        { type: 'null' },
        historyInterruptSchema,
        evidenceSelectionInterruptSchema,
        reviewInterruptSchema,
      ],
    },
    draft_report: {
      anyOf: [
        { type: 'null' },
        draftReportSchema,
      ],
    },
    search_results: { type: 'array', items: searchResultSchema },
    selected_evidence_ids: { type: 'array', items: { type: 'string' } },
    selected_evidence: { type: 'array', items: selectedEvidenceSchema },
    final_report: {
      anyOf: [
        { type: 'null' },
        finalReportSchema,
      ],
    },
    reused_topic: {
      anyOf: [
        { type: 'null' },
        pastTopicSchema,
      ],
    },
  },
}

export const translateTextRequestSchema = {
  type: 'object',
  additionalProperties: false,
  required: ['text', 'target_language'],
  properties: {
    text: { type: 'string', minLength: 1, maxLength: 3200 },
    target_language: { type: 'string', minLength: 2, maxLength: 16 },
  },
}

export const translateTextResponseSchema = {
  type: 'object',
  additionalProperties: false,
  required: ['translated_text', 'target_language'],
  properties: {
    translated_text: nonEmptyString,
    target_language: { type: 'string', minLength: 2, maxLength: 16 },
  },
}
